### PR TITLE
Jic191278

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/kata1.fizzBuzz.js [hunk 1]
```diff
@@ -1,3 +1,10 @@
-const fizzBuzz = (number) => {};
+const fizzBuzz = (number) => {
+	const divBy3 = number % 3 === 0;
+	const divBy5 = number % 5 === 0;
+	if (divBy3 && divBy5) return 'FizzBuzz';
+	if (divBy3) return 'Fizz';
+	if (divBy5) return 'Buzz';
+	return number;
```
Comment (PerryBaran) [inline src/kata1.fizzBuzz.js:7]: good logic, but you should space this out more to improve readability. There is set way to space out your code, but generally you should try to leave gaps between different jobs/things your trying to do, e.g.
```
const fizzBuzz = (number) => {
  const divBy3 = number % 3 === 0;
  const divBy5 = number % 5 === 0;

  if (divBy3 && divBy5) return 'FizzBuzz';

  if (divBy3) return 'Fizz';

  if (divBy5) return 'Buzz';

  return number;
};
```

File: test/kata1.fizzBuzz.test.js [hunk 2]
```diff
@@ -1,11 +1,25 @@
...
+	});
 
-  test("returns the number when it isn't a multiple of 3 or 5", () => {});
+	test("returns the number when it isn't a multiple of 3 or 5", () => {
+		expect(fizzBuzz(4)).toBe(4);
+		expect(fizzBuzz(8)).toBe(8);
+	});
 });
+
+// npm run test -- fizzBuzz
```
Comment (PerryBaran) [inline test/kata1.fizzBuzz.test.js:25]: Should remove this comment

File: src/kata2.booleanToWord.js [hunk 3]
```diff
@@ -1,3 +1,5 @@
-const booleanToWord = (boolean) => {};
+const booleanToWord = (boolean) => {
+	return boolean ? 'Yes' : 'No';
+};
```
Comment (PerryBaran) [inline src/kata2.booleanToWord.js:3]: clean solution

File: test/kata2.booleanToWord.test.js [hunk 4]
```diff
@@ -1,5 +1,8 @@
-const { booleanToWord } = require("../src");
+const { booleanToWord } = require('../src');
 
-describe("booleanToWord", () => {
-  // how do we create specs again???
+describe('booleanToWord', () => {
+	test('check boolean returns correct string', () => {
+		expect(booleanToWord(true)).toBe('Yes');
+		expect(booleanToWord(false)).toBe('No');
+	});
```
Comment (PerryBaran) [inline test/kata2.booleanToWord.test.js:7]: I would separated out these expect statements into separate tests blocks. It improves readability and, if the function failed, would make it more obvious why it's failing. e.g.
```
test("returns 'Yes' when passed true", () => {
  expect(booleanToWord(true)).toBe("Yes");
});

test("returns 'No' when passed false", () => {
  expect(booleanToWord(false)).toBe("No");
});
```

File: src/kata3.numberToReversedDigits.js [hunk 5]
```diff
@@ -1,3 +1,9 @@
-const numberToReversedDigits = (number) => {};
+const numberToReversedDigits = (number) => {
+	return number
+		.toString()
+		.split('')
+		.reverse()
+		.map((item) => Number(item));
+};
```
Comment (PerryBaran) [inline src/kata3.numberToReversedDigits.js:7]: Good solution

File: test/kata4.humanCatDogYears.test.js [hunk 6]
```diff
@@ -1,3 +1,11 @@
...
 
-// Look Ma, no handlebars!!!
+describe('humanCatDogYears', () => {
+	test('check number inputs', () => {
+		expect(humanCatDogYears(0)).toStrictEqual([0, 0, 0]);
+		expect(humanCatDogYears(1)).toStrictEqual([1, 15, 15]);
+		expect(humanCatDogYears(2)).toStrictEqual([2, 24, 24]);
+		expect(humanCatDogYears(3)).toStrictEqual([3, 28, 29]);
+		expect(humanCatDogYears(4)).toStrictEqual([4, 32, 34]);
+	});
```
Comment (PerryBaran) [inline test/kata4.humanCatDogYears.test.js:10]: as previously mentioned, it would be better separate out each of these expects into their own test block with a descriptive test description for each

File: src/kata5.reachDestination.js [hunk 7]
```diff
@@ -1,3 +1,5 @@
-const reachDestination = (distance, speed) => {};
+const reachDestination = (distance, speed) => {
+	return `I should be there in ${Math.ceil((distance / speed) * 2) / 2} hours`;
```
Comment (PerryBaran) [inline src/kata5.reachDestination.js:2]: The returned string is grammatically incorrect if the arrival time is 1. i.e. it should be `"I should be there in 1 hour"` but currently it returns `"I should be there in 1 hours"`.

To fix this, and also improve readability, we can take out the arrival time calculation and assign it to a variable. i.e.
```
const arrival = Math.ceil((distance / speed) * 2) / 2
return `I should be there in ${arrival} hours`;
```
then, we can use this arrival time to conditionally add the `'s'` at the end of `'hour'`, e.g.
```
const arrival = Math.ceil((distance / speed) * 2) / 2
return `I should be there in ${arrival} hour${arrival === 1 ? "" : "s"}`;
```

File: test/kata5.reachDestination.test.js [hunk 8]
```diff
@@ -1,5 +1,7 @@
-const { reachDestination } = require("../src");
+const { reachDestination } = require('../src');
 
-describe("reachDestination", () => {
-  test("returns string with estimated time of arrival", () => {});
+describe('reachDestination', () => {
+	test('returns string with estimated time of arrival', () => {
+		expect(reachDestination(44, 10)).toBe('I should be there in 4.5 hours');
+	});
```
Comment (PerryBaran) [inline test/kata5.reachDestination.test.js:6]: You should add another test for the edge case when arrival time is 1, to check that the output is as expected, e.g.
```
test("returns string with estimated time of arrival when arrival time is 1", () => {
    expect(reachDestination(1, 1)).toBe("I should be there in 1 hour");
  });
```

File: src/kata6.joinNames.js [hunk 9]
```diff
@@ -1,3 +1,26 @@
...
+// for (let i = 1; i < namesObj.length; i++) {
+// 	if (namesObj.length === 1) break;
+// 	if (i !== namesObj.length - 1) {
+// 		string += `, ${namesObj[i].name}`;
+// 	} else {
+// 		string += ` & ${namesObj[i].name}`;
+// 	}
+// }
+// console.log(string);
+// return string;
```
Comment (PerryBaran) [inline src/kata6.joinNames.js:26]: Should remove commented out code after you are done with it

File: src/kata6.joinNames.js [hunk 10]
```diff
@@ -1,3 +1,26 @@
-const joinNames = (namesObj) => {};
+const joinNames = (namesObj) => {
+	if (namesObj.length === 0) return '';
+	const names = namesObj.map((name) => name.name);
+	const lastName = names.pop();
+	console.log(names, lastName);
```
Comment (PerryBaran) [inline src/kata6.joinNames.js:5]: Should remove console.logs after you are done with them

File: test/kata6.joinNames.test.js [hunk 11]
```diff
@@ -1,5 +1,21 @@
...
+		).toBe('Bart, Lisa & Maggie');
+		expect(
+			joinNames([
+				{ name: 'Bart' },
+				{ name: 'Lisa' },
+				{ name: 'Maggie' },
+				{ name: 'Homer' },
+				{ name: 'Marge' },
+			])
+		).toBe('Bart, Lisa, Maggie, Homer & Marge');
```
Comment (PerryBaran) [inline test/kata6.joinNames.test.js:19]: I like that you've added tests for different array lengths. However, it would improve readability if you separated these out into their own test blocks. e.g.
```
 test("returns an empty string for empty arrays", () => {
    expect(joinNames([])).toBe("");
  });

  test("returns string of name for array of length 1", () => {
    expect(joinNames([{ name: "Bart" }])).toBe("Bart");
  });

  test("returns string names, separated by ampersand for array of length 2", () => {
    expect(joinNames([{ name: "Bart" }, { name: "Lisa" }])).toBe("Bart & Lisa");
  });

 test("returns string names, separated by ampersand for array of length 3", () => {
    expect(joinNames([{ name: "Bart" }, { name: "Lisa" }, { name: "Maggie" }])).toBe("Bart, Lisa & Maggie");
  });

  test("returns string of names, separated by commas and an ampersand, for long arrays", () => {
    expect(
      joinNames([
        {
          name: "Bart",
        },
        {
          name: "Lisa",
        },
        {
          name: "Maggie",
        },
        {
          name: "Homer",
        },
        {
          name: "Marge",
        },
      ])
    ).toBe("Bart, Lisa, Maggie, Homer & Marge");
  });
```

File: test/kata6.joinNames.test.js [hunk 12]
```diff
@@ -1,5 +1,21 @@
-const { joinNames } = require("../src");
+const { joinNames } = require('../src');
 
-describe("joinNames", () => {
-  test("returns string of names, seperated by commas and an ampersand", () => {});
+describe('joinNames', () => {
+	test('returns string of names, seperated by commas and an ampersand', () => {
```
Comment (PerryBaran) [inline test/kata6.joinNames.test.js:4]: spelling mistake here, should be `separated` instead of `seperated`

File: src/kata7.getEmployerRole.js [hunk 13]
```diff
@@ -1,3 +1,14 @@
...
+	const empNames = employees.filter((emp) => employeeName === emp.name);
+	if (empNames.length === 0) return `${employeeName} does not work here!`;
+	let string = '';
+	for (let i = 0; i < empNames.length; i++) {
+		i === empNames.length - 1
+			? (string += empNames[i].role)
+			: (string += `${empNames[i].role} `);
+	}
+	return string;
+};
```
Comment (PerryBaran) [inline src/kata7.getEmployerRole.js:12]: Interesting take on this problem. Usually, you just return the first role of the employee that has that name, e.g.
```
const getEmployerRole = (employeeName, employees) => {
  return employees.find(employee => employee.name === employeeName).role;
}
```
However, you've set this up to return all roles of an employee that have that name

File: test/kata7.getEmployerRole.test.js [hunk 14]
```diff
@@ -1,5 +1,35 @@
...
+			'Human Recommended Reading Assistant'
+		);
+	});
+	test('check duplicates', () => {
+		employees.push({ name: 'Satti', role: 'Admin' });
+		expect(getEmployerRole('Satti', employees)).toBe('Developer Admin');
+		employees.push({ name: 'Satti', role: 'Cleaner' });
+		expect(getEmployerRole('Satti', employees)).toBe('Developer Admin Cleaner');
+	});
 });
```
Comment (PerryBaran) [inline test/kata7.getEmployerRole.test.js:35]: Having one `employees` array that is used by each test that you keep adding to (`.push`) is bad practice. It could lead to unexpected problems down the line. For now, it would be better to define the `employees` array in each test with it's own variables. However, this will mean you'll have a lot of code repetition. Further into the track we will go over some ways you can reduce this repetition to make your code more DRY.

**Review by PerryBaran — COMMENTED** [review]
Overall some good logic. I like that you've added tests for different setups. 

One thing I haven't mentioned elsewhere is that your code has a lot of formatting issues. If you have red squiggly lines in your code editor and your files names show up in red on the side, you need to fix these issues. If you hover your mouse over the red squiggly line it will give you information on what is causing the problem. At the bottom of this box is the text Quick Fix (Ctrl+.) (different symbol to Ctrl on mac). Clicking this button will bring up another popup with suggestions on how to fix/workaround this problem. The first option in this popup will usually fix your problem for you.
